geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AccumuloJoinIndex.scala

```diff
@@ -0,0 +1,395 @@
+import org.apache.accumulo.core.data.{Key, Range, Value}
+import org.apache.hadoop.io.Text
+import org.geotools.factory.Hints
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder
+import org.locationtech.geomesa.accumulo.data.AccumuloQueryPlan.{BatchScanPlan, JoinFunction, JoinPlan}
+import org.locationtech.geomesa.accumulo.data.{AccumuloDataStore, AccumuloIndexAdapter, AccumuloQueryPlan}
+import org.locationtech.geomesa.accumulo.iterators._
+import org.locationtech.geomesa.filter.{FilterHelper, andOption, partitionPrimarySpatials, partitionPrimaryTemporals}
+import org.locationtech.geomesa.index.api._
+import org.locationtech.geomesa.index.index.attribute.{AttributeIndex, AttributeIndexKey, AttributeIndexValues}
+import org.locationtech.geomesa.index.index.id.IdIndex
+import org.locationtech.geomesa.index.iterators.StatsScan
+import org.locationtech.geomesa.index.planning.LocalQueryRunner
+import org.locationtech.geomesa.index.planning.LocalQueryRunner.ArrowDictionaryHook
+import org.locationtech.geomesa.index.stats.GeoMesaStats
+import org.locationtech.geomesa.utils.collection.CloseableIterator
+import org.locationtech.geomesa.utils.index.{ByteArrays, IndexMode, VisibilityLevel}
+import org.locationtech.geomesa.utils.stats.{Cardinality, Stat}
+import org.opengis.feature.`type`.AttributeDescriptor
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter.Filter
+
+import scala.util.Try
+
+/**
+  * Mixin trait to add join support to the normal attribute index class
+  */
+trait AccumuloJoinIndex extends GeoMesaFeatureIndex[AttributeIndexValues[Any], AttributeIndexKey] {
+
+  this: AttributeIndex =>
+
+  import org.locationtech.geomesa.index.conf.QueryHints.RichHints
+  import org.locationtech.geomesa.utils.geotools.RichAttributeDescriptors.RichAttributeDescriptor
+  import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+
+  import scala.collection.JavaConverters._
+
+  protected val attributeIndex: Int = sft.indexOf(attribute)
+  protected val descriptor: AttributeDescriptor = sft.getDescriptor(attributeIndex)
+  protected val binding: Class[_] = descriptor.getType.getBinding
+  protected val indexSft: SimpleFeatureType = IndexValueEncoder.getIndexSft(sft)
+
+  override val name: String = JoinIndex.name
+  override val identifier: String = GeoMesaFeatureIndex.identifier(name, version, attributes)
+
+  abstract override def getFilterStrategy(filter: Filter,
+                                          transform: Option[SimpleFeatureType],
+                                          stats: Option[GeoMesaStats]): Option[FilterStrategy] = {
+    super.getFilterStrategy(filter, transform, stats).flatMap { strategy =>
+      val join = requiresJoin(strategy.secondary, transform)
+      // verify that it's ok to return join plans, and filter them out if not
+      if (join && !JoinIndex.AllowJoinPlans.get) { None } else {
```

Comment:
The naked call to .get on an Option makes me nervous:).

Reply:
that's actually a thread-local, not an option :)

Reply:
with a default value!